src/utilities/geometry/Polyhedron.cpp

```diff
@@ -58,136 +67,246 @@ bool Surface3dEdge::operator!=(const Surface3dEdge& other) const {
+    }
 
-size_t Polyhedron::numVertices() const {
-  size_t count = 0;
-  for (const auto& surface : m_surfaces) {
-    count += surface.vertices.size();
+    edges.emplace_back(*it, *itnext, t_name, t_surfNum);
   }
-  return count;
 }
 
-std::vector<Point3d> Polyhedron::uniqueVertices() const {
+bool Surface3d::operator<(const Surface3d& rhs) const {
+  return this->name < rhs.name;
+}
 
-  std::vector<Point3d> uniqVertices;
-  uniqVertices.reserve(numVertices());
+size_t Surface3d::numConflictedEdges() const {
+  return std::count_if(edges.cbegin(), edges.cend(), [](const auto& edge) { return edge.hasConflictedOrientation(); });
+}
 
-  for (const auto& surface : m_surfaces) {
-    for (const auto& pt : surface.vertices) {
-      if (std::find_if(uniqVertices.cbegin(), uniqVertices.cend(), [&pt](const auto& unqV) { return isAlmostEqual3dPt(pt, unqV); })
-          == uniqVertices.cend()) {
-        uniqVertices.push_back(pt);
+double Surface3d::ratioOfConflictedEdges() const {
+  return static_cast<double>(numConflictedEdges()) / static_cast<double>(edges.size());
+}
+
+void Surface3d::resetEdgeMatching() {
+  for (auto& edge : edges) {
+    edge.resetEdgeMatching();
+  }
+}
+
+bool Surface3d::isConvex() const {
+
+  // const auto& a = edges.front().start();
+  // const auto& b = edges.front().end();
+  // const auto& c = edges.at(1).end();
+
+  auto ab = edges.front().asVector();
+  auto bc = edges[1].asVector();
+  auto outwardNormal = ab.cross(bc);
+  outwardNormal.normalize();
+
+  for (auto it = std::next(edges.begin()); it != edges.end(); ++it) {
+    auto itnext = std::next(it);
+    if (itnext == std::end(edges)) {
+      itnext = std::begin(edges);
+    }
+    const Point3d& a = it->start();
+    const Point3d& b = it->end();
+    const Point3d& c = itnext->end();
+
+    auto ab = b - a;
+    auto ac = c - a;
+
+    auto triangleNormal = ab.cross(ac);
+    auto d = outwardNormal.dot(triangleNormal);
+    if (d < 0) {
+      return false;
+    }
+  }
+  return true;
+}
+
+Polyhedron::Polyhedron(std::vector<Surface3d> surfaces) : m_surfaces(std::move(surfaces)) {
+  performEdgeMatching();
+
+  // if all edges had two counts then it is fully enclosed
+  if (!hasEdgesNot2()) {
+    m_isEnclosedVolume = true;
+  } else {
+    LOG(Warn, "Polyhedron is not enclosed in original testing. Trying to add missing colinear points.");
+    updateZonePolygonsForMissingColinearPoints();
+  }
+  if (!m_isEnclosedVolume) {
+    LOG(Warn, "Polyhedron is not enclosed.");
+  } else {
+    // We call the volume calculation... if this ends up being negative, we're in the case where ALL surfaces are in the wrong orientation
+    if (!m_hasAnySurfaceWithIncorrectOrientation) {
+      m_polyhedronVolume = calcPolyhedronVolume();
+      if (m_polyhedronVolume < 0) {
+        LOG(Error, "It seems that ALL surfaces are reversed");
+        m_hasAnySurfaceWithIncorrectOrientation = true;
+        m_isCompletelyInsideOut = true;
```

Comment:
Catch the case where ALL surfaces are inside out.... If we initially didn't find any orientation issues, but the Polyhedron's volume is negative, then it is inside out.